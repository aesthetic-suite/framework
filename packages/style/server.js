module.exports = require('./lib/server/index.js');
