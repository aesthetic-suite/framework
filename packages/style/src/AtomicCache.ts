import { CacheItem, StyleParams } from './types';

// https://jsperf.com/javascript-objects-vs-map-performance
// https://jsperf.com/performance-of-map-vs-object
export default class AtomicCache {
  cache: { [property: string]: { [value: string]: CacheItem[] } } = {};

  match(item: CacheItem, params: StyleParams, minimumRank?: number): boolean {
    if (minimumRank !== undefined && item.rank < minimumRank) {
      return false;
    }

    if (item.selector !== params.selector) {
      return false;
    }

    if (item.type !== params.type) {
      return false;
    }

    if (Array.isArray(params.conditions) && params.conditions.length > 0) {
      return params.conditions.every(i =>
        item.conditions.find(p => p.query === i.query && p.type === i.type),
      );
    }

    return true;
  }

  read(
    property: string,
    value: string,
    params: StyleParams,
    minimumRank?: number,
  ): CacheItem | null {
    const propertyCache = this.cache[property];

    if (!propertyCache) {
      return null;
    }

    const valueCache = propertyCache[value];

    if (!valueCache || valueCache.length === 0) {
      return null;
    }

    return valueCache.find(item => this.match(item, params, minimumRank)) ?? null;
  }

  write(property: string, value: string, item: CacheItem): this {
    let propertyCache = this.cache[property];

    if (!propertyCache) {
      propertyCache = {};
      this.cache[property] = propertyCache;
    }

    let valueCache = propertyCache[value];

    if (!valueCache) {
      valueCache = [];
      propertyCache[value] = valueCache;
    }

    valueCache.push(item);

    return this;
  }
}
