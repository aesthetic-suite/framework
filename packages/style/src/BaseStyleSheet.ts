import { isSSR } from '@aesthetic/utils';
import { StyleRule } from './types';

export default abstract class BaseStyleSheet {
  lastIndex: number = -1;

  sheet: StyleRule;

  protected rafHandle: number = 0;

  protected ruleBuffer: { index: number; rule: string }[] = [];

  constructor(sheet: StyleRule) {
    this.sheet = sheet;
  }

  protected enqueueRule(rule: string, customIndex?: number): number {
    let index = 0;

    // When testing or SSR, just insert the rule immediately
    if (process.env.NODE_ENV === 'test' || isSSR()) {
      index = customIndex ?? this.sheet.cssRules.length;

      this.injectRule(rule, index);

      // Otherwise buffer the insert until the next animation frame
    } else {
      index = customIndex ?? (this.lastIndex += 1);

      this.ruleBuffer.push({
        index,
        rule,
      });

      if (!this.rafHandle) {
        this.rafHandle = requestAnimationFrame(this.flushRules);
      }
    }

    return index;
  }

  protected injectRule = (rule: string, index: number) => {
    // CSS is locked in developer tools when using `insertRule`,
    // so avoid using it in development so we have full control.
    // istanbul ignore next
    if (process.env.NODE_ENV === 'development' && !isSSR()) {
      this.sheet.textContent += rule;
    } else {
      this.sheet.insertRule(rule, index);
      this.lastIndex = this.sheet.cssRules.length;
    }
  };

  protected flushRules = () => {
    this.rafHandle = 0;
    this.ruleBuffer.forEach((buffer) => {
      this.injectRule(buffer.rule, buffer.index);
    });
  };
}
