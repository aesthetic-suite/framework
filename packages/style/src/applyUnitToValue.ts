import { isUnitlessProperty } from 'css-in-js-utils';
import { Value } from './types';

export default function applyUnitToValue(property: string, value: Value): string {
  if (typeof value === 'string' || isUnitlessProperty(property) || value === 0) {
    return String(value);
  }

  return `${value}px`;
}
