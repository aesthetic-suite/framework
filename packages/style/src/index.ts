/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import { Direction, EngineOptions, VariablesMap } from '@aesthetic/types';
import { objectLoop } from '@aesthetic/utils';
import { createStyleElements, getStyleElement } from './client/dom';
import { hydrateStyles } from './client/hydration';
import { createCacheManager } from './common/cache';
import { createStyleEngine } from './common/engine';
import { createSheetManager } from './common/sheet';
import { formatVariable } from './common/syntax';
import { StyleEngine } from './types';

export * from './common/cache';
export * from './common/constants';
export * from './common/engine';
export * from './common/helpers';
export * from './common/sheet';
export * from './common/syntax';
export * from './types';

export { createStyleElements, getStyleElement };

// Set active direction on the document `dir` attribute
function setDirection(direction: Direction) {
  document.documentElement.setAttribute('dir', direction);
}

// Set CSS variables to :root
function setRootVariables(vars: VariablesMap) {
  objectLoop(vars, (value, key) => {
    document.documentElement.style.setProperty(formatVariable(key), String(value));
  });
}

// Set active theme class names on the `body`
function setTheme(classNames: string[]) {
  document.body.className = classNames.join(' ');
}

export function createClientEngine(options: Partial<EngineOptions> = {}): StyleEngine {
  if (global.AESTHETIC_CUSTOM_ENGINE) {
    return global.AESTHETIC_CUSTOM_ENGINE as StyleEngine;
  }

  const direction = (document.documentElement.getAttribute('dir') ||
    document.body.getAttribute('dir') ||
    'ltr') as Direction;

  const engine: StyleEngine = createStyleEngine({
    cacheManager: createCacheManager(),
    direction,
    sheetManager: createSheetManager(createStyleElements()),
    ...options,
  });

  // Match against browser preferences
  engine.prefersColorScheme = (scheme) => matchMedia(`(prefers-color-scheme: ${scheme})`).matches;
  engine.prefersContrastLevel = (level) => matchMedia(`(prefers-contrast: ${level})`).matches;

  // Handle DOM specific logic
  engine.setDirection = setDirection;
  engine.setRootVariables = setRootVariables;
  engine.setTheme = setTheme;

  // Attempt to hydrate styles immediately
  hydrateStyles(engine);

  return engine;
}
