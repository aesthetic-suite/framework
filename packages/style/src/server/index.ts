import { Variables } from '@aesthetic/types';
import { objectLoop } from '@aesthetic/utils';
import { TransientSheet, sortConditionalRules } from './sheet';
import {
  createCacheManager,
  createStyleEngine,
  createSheetManager,
  formatVariable,
  EngineOptions,
  StyleEngine,
} from '../index';

function setRootVariables(vars: Variables, engine: StyleEngine) {
  const sheet = engine.sheetManager.sheets.global;

  objectLoop(vars, (value, key) => {
    sheet.cssVariables[formatVariable(key)] = String(value);
  });
}

export function createServerEngine(options: Partial<EngineOptions>): StyleEngine {
  const sheetManager = createSheetManager({
    conditions: new TransientSheet(),
    global: new TransientSheet(),
    standard: new TransientSheet(),
  });

  const engine: StyleEngine = {
    ...createStyleEngine({
      cacheManager: createCacheManager(),
      sheetManager: {
        ...sheetManager,
        insertRule(type, rule, index) {
          const rank = sheetManager.insertRule(type, rule, index);

          if (type === 'conditions') {
            sortConditionalRules(sheetManager.sheets.conditions);
          }

          return rank;
        },
      },
      ...options,
    }),
    setRootVariables: (vars) => setRootVariables(vars, engine),
  };

  return engine;
}

export function extractStyles<T>(result: T, engine: StyleEngine): T {
  // global.AESTHETIC_CUSTOM_RENDERER = engine;
  process.env.AESTHETIC_SSR = 'true';

  return result;
}
