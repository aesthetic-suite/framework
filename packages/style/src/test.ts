import {
	CacheManager,
	ClassName,
	CSS,
	EngineOptions,
	Sheet,
	SheetManager,
	SheetType,
} from '@aesthetic/types';
// Rollup compatibility
import {
	createCacheManager,
	createSheetManager,
	createStyleElements,
	createStyleEngine,
	getStyleElement,
	StyleEngine,
} from '.';

export function createTestCacheManager(): CacheManager<ClassName> {
	return createCacheManager();
}

export function createTestSheetManager(): SheetManager {
	return createSheetManager(createStyleElements());
}

export function createTestStyleEngine(
	options: Partial<EngineOptions<ClassName>> = {},
): StyleEngine {
	return createStyleEngine({
		cacheManager: createTestCacheManager(),
		sheetManager: createTestSheetManager(),
		...options,
	});
}

export function getRenderedStyles(type: Sheet | SheetType): CSS {
	return [...(typeof type === 'string' ? getStyleElement(type) : type).cssRules].reduce(
		(css, rule) => css + rule.cssText,
		'',
	);
}

export function purgeStyles(type?: SheetType): void {
	if (type) {
		// This is the only way to generate accurate snapshots.
		// It may slow down tests though?
		document.querySelector(`#aesthetic-${type}`)?.remove();
	} else {
		purgeStyles('global');
		purgeStyles('standard');
		purgeStyles('conditions');
	}
}
