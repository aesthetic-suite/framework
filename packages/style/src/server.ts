import { EngineOptions, Variables } from '@aesthetic/types';
import { objectLoop } from '@aesthetic/utils';
import { TransientSheet, createSheetManager } from './server/sheet';
import { createCacheManager, createStyleEngine, formatVariable, StyleEngine } from './index';

function setRootVariables(vars: Variables, engine: StyleEngine) {
  const sheet = engine.sheetManager.sheets.global;

  objectLoop(vars, (value, key) => {
    sheet.cssVariables[formatVariable(key)] = String(value);
  });
}

export function createServerEngine(options: Partial<EngineOptions>): StyleEngine {
  const engine: StyleEngine = {
    ...createStyleEngine({
      cacheManager: createCacheManager(),
      sheetManager: createSheetManager({
        conditions: new TransientSheet(),
        global: new TransientSheet(),
        standard: new TransientSheet(),
      }),
      ...options,
    }),
    setRootVariables: (vars) => setRootVariables(vars, engine),
  };

  return engine;
}

export function extractStyles<T>(result: T, engine: StyleEngine): T {
  global.AESTHETIC_CUSTOM_ENGINE = engine;
  process.env.AESTHETIC_SSR = 'true';

  return result;
}
