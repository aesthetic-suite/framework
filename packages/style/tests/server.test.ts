import { ServerStyleEngine } from '../src';
import { createServerEngine, renderToStyleMarkup } from '../src/server';
import { purgeStyles } from '../src/test';

describe('Server', () => {
	let engine: ServerStyleEngine;

	beforeEach(() => {
		engine = createServerEngine();

		global.AESTHETIC_CUSTOM_ENGINE = engine;
		process.env.AESTHETIC_SSR = 'true';
	});

	afterEach(() => {
		purgeStyles();

		// @ts-expect-error Allow delete
		delete global.AESTHETIC_CUSTOM_ENGINE;
	});

	it('sets SSR global', () => {
		// @ts-expect-error Allow delete
		delete global.AESTHETIC_CUSTOM_ENGINE;

		engine.extractStyles();

		expect(global.AESTHETIC_CUSTOM_ENGINE).toBe(engine);
	});

	it('sets SSR env var', () => {
		delete process.env.AESTHETIC_SSR;

		engine.extractStyles();

		expect(process.env.AESTHETIC_SSR).toBe('true');
	});

	it('writes to a temporary style sheet implementation and generates accurate markup', () => {
		engine.setRootVariables({
			fontSize: '16px',
			bgColor: '#fff',
			fbColor: 'black',
		});

		engine.renderRule({
			margin: 0,
			padding: '6px 12px',
			border: '1px solid #2e6da4',
			borderRadius: '4px',
			display: 'inline-block',
			cursor: 'pointer',
			fontFamily: 'Roboto',
			fontWeight: 'normal',
			lineHeight: 'normal',
			whiteSpace: 'nowrap',
			textDecoration: 'none',
			textAlign: 'left',
			backgroundColor: '#337ab7',
			verticalAlign: 'middle',
			color: 'rgba(0, 0, 0, 0)',
			animationName: 'fade',
			animationDuration: '.3s',
		});

		engine.renderFontFace({
			fontFamily: '"Open Sans"',
			fontStyle: 'normal',
			fontWeight: 800,
			src: 'url("fonts/OpenSans-Bold.woff2")',
		});

		engine.renderRule({
			margin: 0,
			'@media': {
				'(width: 500px)': {
					margin: '10px',
					':hover': {
						color: 'red',
					},
					'@media': {
						'(width: 350px)': {
							'@supports': {
								'(color: blue)': {
									color: 'blue',
								},
							},
						},
					},
				},
			},
			'@supports': {
				'(color: green)': {
					color: 'green',
				},
			},
		});

		engine.renderKeyframes({
			from: {
				transform: 'translateX(0%)',
			},
			to: {
				transform: 'translateX(100%)',
			},
		});

		engine.renderImport({ path: 'test.css', url: true });

		// Test that conditions are merged
		engine.renderRule({
			margin: 0,
			'@media': {
				'(width: 500px)': {
					margin: '5px',
				},
			},
			'@supports': {
				'(color: green)': {
					color: 'black',
				},
			},
		});

		expect(renderToStyleMarkup(engine)).toMatchSnapshot();
	});

	it('can render media and feature queries', () => {
		engine.renderRule({
			'@media': {
				'(max-width: 1000px)': { display: 'block' },
			},
			'@supports': {
				'(display: flex)': { display: 'flex' },
			},
		});

		expect(renderToStyleMarkup(engine)).toMatchSnapshot();
	});

	it('merges conditionals instead of appending', () => {
		engine.renderRule({
			'@media': { '(max-width: 1000px)': { display: 'block' } },
		});
		engine.renderRule({
			'@media': { '(max-width: 1000px)': { display: 'inline-block' } },
		});
		engine.renderRule({
			'@media': { '(max-width: 1000px)': { display: 'flex' } },
		});

		expect(renderToStyleMarkup(engine)).toMatchSnapshot();
	});

	it('can render CSS variables', () => {
		engine.setRootVariables({
			'--root-level': 'true',
		});

		engine.renderRule({
			'@variables': {
				'--element-level': 123,
			},
		});

		engine.renderVariable('varLevel', '10px');

		expect(renderToStyleMarkup(engine)).toMatchSnapshot();
	});

	it('sorts media queries using mobile-first', () => {
		const block = { padding: 0 } as const;

		engine.renderRule({
			'@media': {
				'screen and (min-width: 1024px)': block,
				'screen and (min-width: 320px) and (max-width: 767px)': block,
				'screen and (min-width: 1280px)': block,
				'screen and (min-height: 480px)': block,
				'screen and (min-height: 480px) and (min-width: 320px)': block,
				'screen and (orientation: portrait)': block,
				'screen and (min-width: 640px)': block,
				print: block,
				'screen and (max-width: 767px) and (min-width: 320px)': block,
				tv: block,
				'screen and (max-height: 767px) and (min-height: 320px)': block,
				'screen and (orientation: landscape)': block,
				'screen and (min-device-width: 320px) and (max-device-width: 767px)': block,
				'screen and (max-width: 639px)': block,
				'screen and (max-width: 1023px)': block,
			},
		});

		expect(renderToStyleMarkup(engine)).toMatchSnapshot();
	});
});
