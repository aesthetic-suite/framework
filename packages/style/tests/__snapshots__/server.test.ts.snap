// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server can render CSS variables 1`] = `"<style  data-aesthetic-hydrate-index=\\"-1\\" data-aesthetic-rule-count=\\"1\\" data-aesthetic-type=\\"global\\" id=\\"aesthetic-global\\" media=\\"screen\\" type=\\"text/css\\">:root { --root-level:true; }</style><style  data-aesthetic-hydrate-index=\\"1\\" data-aesthetic-rule-count=\\"1\\" data-aesthetic-type=\\"standard\\" id=\\"aesthetic-standard\\" media=\\"screen\\" type=\\"text/css\\">.a { --element-level:123; }.b { --var-level:10px; }</style><style  data-aesthetic-hydrate-index=\\"-1\\" data-aesthetic-rule-count=\\"1\\" data-aesthetic-type=\\"conditions\\" id=\\"aesthetic-conditions\\" media=\\"screen\\" type=\\"text/css\\"></style>"`;

exports[`Server can render media and feature queries 1`] = `"<style  data-aesthetic-hydrate-index=\\"-1\\" data-aesthetic-rule-count=\\"1\\" data-aesthetic-type=\\"global\\" id=\\"aesthetic-global\\" media=\\"screen\\" type=\\"text/css\\"></style><style  data-aesthetic-hydrate-index=\\"-1\\" data-aesthetic-rule-count=\\"1\\" data-aesthetic-type=\\"standard\\" id=\\"aesthetic-standard\\" media=\\"screen\\" type=\\"text/css\\"></style><style  data-aesthetic-hydrate-index=\\"1\\" data-aesthetic-rule-count=\\"1\\" data-aesthetic-type=\\"conditions\\" id=\\"aesthetic-conditions\\" media=\\"screen\\" type=\\"text/css\\">@media (max-width: 1000px) { .a { display:block; } }@supports (display: flex) { .b { display:flex; } }</style>"`;

exports[`Server merges conditionals instead of appending 1`] = `"<style  data-aesthetic-hydrate-index=\\"-1\\" data-aesthetic-rule-count=\\"2\\" data-aesthetic-type=\\"global\\" id=\\"aesthetic-global\\" media=\\"screen\\" type=\\"text/css\\"></style><style  data-aesthetic-hydrate-index=\\"-1\\" data-aesthetic-rule-count=\\"2\\" data-aesthetic-type=\\"standard\\" id=\\"aesthetic-standard\\" media=\\"screen\\" type=\\"text/css\\"></style><style  data-aesthetic-hydrate-index=\\"0\\" data-aesthetic-rule-count=\\"2\\" data-aesthetic-type=\\"conditions\\" id=\\"aesthetic-conditions\\" media=\\"screen\\" type=\\"text/css\\">@media (max-width: 1000px) { .a { display:block; }.b { display:inline-block; }.c { display:flex; } }</style>"`;

exports[`Server sorts media queries using mobile-first 1`] = `"<style  data-aesthetic-hydrate-index=\\"-1\\" data-aesthetic-rule-count=\\"14\\" data-aesthetic-type=\\"global\\" id=\\"aesthetic-global\\" media=\\"screen\\" type=\\"text/css\\"></style><style  data-aesthetic-hydrate-index=\\"-1\\" data-aesthetic-rule-count=\\"14\\" data-aesthetic-type=\\"standard\\" id=\\"aesthetic-standard\\" media=\\"screen\\" type=\\"text/css\\"></style><style  data-aesthetic-hydrate-index=\\"14\\" data-aesthetic-rule-count=\\"14\\" data-aesthetic-type=\\"conditions\\" id=\\"aesthetic-conditions\\" media=\\"screen\\" type=\\"text/css\\">@media screen and (min-width: 320px) and (max-width: 767px) { .b { padding:0; } }@media screen and (min-height: 480px) { .d { padding:0; } }@media screen and (min-height: 480px) and (min-width: 320px) { .e { padding:0; } }@media screen and (min-width: 640px) { .g { padding:0; } }@media screen and (min-width: 1024px) { .a { padding:0; } }@media screen and (min-width: 1280px) { .c { padding:0; } }@media screen and (min-device-width: 320px) and (max-device-width: 767px) { .m { padding:0; } }@media screen and (max-width: 1023px) { .o { padding:0; } }@media screen and (max-height: 767px) and (min-height: 320px) { .k { padding:0; } }@media screen and (max-width: 767px) and (min-width: 320px) { .i { padding:0; } }@media screen and (max-width: 639px) { .n { padding:0; } }@media screen and (orientation: landscape) { .l { padding:0; } }@media screen and (orientation: portrait) { .f { padding:0; } }@media tv { .j { padding:0; } }@media print { .h { padding:0; } }</style>"`;

exports[`Server writes to a temporary style sheet implementation and generates accurate markup 1`] = `"<style  data-aesthetic-hydrate-index=\\"2\\" data-aesthetic-rule-count=\\"22\\" data-aesthetic-type=\\"global\\" id=\\"aesthetic-global\\" media=\\"screen\\" type=\\"text/css\\">:root { --font-size:16px;--bg-color:#fff;--fb-color:black; }@import url(\\"test.css\\");@font-face { font-family:\\"Open Sans\\";font-style:normal;font-weight:800;src:url(\\"fonts/OpenSans-Bold.woff2\\"); }@keyframes kf103rcyx { from { transform:translateX(0%); } to { transform:translateX(100%); }  }</style><style  data-aesthetic-hydrate-index=\\"16\\" data-aesthetic-rule-count=\\"22\\" data-aesthetic-type=\\"standard\\" id=\\"aesthetic-standard\\" media=\\"screen\\" type=\\"text/css\\">.a { margin:0; }.b { padding:6px 12px; }.c { border:1px solid #2e6da4; }.d { border-radius:4px; }.e { display:inline-block; }.f { cursor:pointer; }.g { font-family:Roboto; }.h { font-weight:normal; }.i { line-height:normal; }.j { white-space:nowrap; }.k { text-decoration:none; }.l { text-align:left; }.m { background-color:#337ab7; }.n { vertical-align:middle; }.o { color:rgba(0, 0, 0, 0); }.p { animation-name:fade; }.q { animation-duration:.3s; }</style><style  data-aesthetic-hydrate-index=\\"2\\" data-aesthetic-rule-count=\\"22\\" data-aesthetic-type=\\"conditions\\" id=\\"aesthetic-conditions\\" media=\\"screen\\" type=\\"text/css\\">@media (width: 500px) { .r { margin:10px; }.t:hover { color:red; }.v { margin:5px; } }@supports (color: blue) { @media (width: 500px) and (width: 350px) { .s { color:blue; } } }@supports (color: green) { .u { color:green; }.w { color:black; } }</style>"`;
